crates/walrus-core/src/encoding/quilt_encoding.rs

```diff
@@ -0,0 +1,1367 @@
+// Copyright (c) Walrus Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+use alloc::{
+    format,
+    string::{String, ToString},
+    vec,
+    vec::Vec,
+};
+use core::{cmp, fmt, num::NonZeroU16};
+
+use crc::{Crc, CRC_8_CDMA2000};
+use hex;
+use serde::{Deserialize, Serialize};
+use tracing::{Level, Span};
+
+use super::{EncodingConfig, EncodingConfigEnum, Primary, Secondary, SliverData, SliverPair};
+use crate::{
+    encoding::{blob_encoding::BlobEncoder, config::EncodingConfigTrait as _, QuiltError},
+    metadata::{QuiltBlock, QuiltIndexV1, QuiltMetadata},
+    SliverIndex,
+};
+
+/// The number of bytes to store the size of the quilt index.
+const QUILT_INDEX_SIZE_PREFIX_SIZE: usize = 8;
+
+/// The maximum number of columns a quilt index can have.
+const MAX_NUM_COLUMNS_FOR_QUILT_INDEX: usize = 1024;
+
+pub trait QuiltVersion {
+    type QuiltConfig;
+    type QuiltEncoder<'a>;
+    type QuiltDecoder<'a>;
+    type Quilt;
+    type QuiltIndex;
+    type QuiltMetadata;
+}
+
+pub trait QuiltConfigApi<'a, V: QuiltVersion> {
+    /// Returns a new encoder for the given configuration and blobs.
+    fn get_encoder(
+        &self,
+        encoding_config: EncodingConfigEnum<'a>,
+        blobs: &'a [BlobWithIdentifier<'a>],
+    ) -> V::QuiltEncoder<'a>;
+
+    /// Returns a new decoder for the given slivers.
+    fn get_decoder(&self, slivers: &'a [&'a SliverData<Secondary>]) -> V::QuiltDecoder<'a>;
+
+    /// Constructs a quilt from a quilt blob.
+    ///
+    /// `quilt_blob` is a quilt constructed from a set of blobs.
+    /// This function loads the quilt blob to access its internal structures without having
+    /// to re-encode the blobs.
+    fn new_from_quilt_blob(
+        &self,
+        quilt_blob: Vec<u8>,
+        metadata: &QuiltMetadata,
```

Comment:
Question: Why do we have to pass in the metadata and where would we get it from?